### PR TITLE
ci(pre-commit): Auto-install clang-format and get rid of wrapper script

### PR DESCRIPTION
**Pull request by Holzhaus**
Fixes the clang-format pre-commit hook on windows.

### CODE REVIEW

**Comment by Holzhaus** [comment]
@JoergAtGithub please have a look and check if this works. It should work even without clang-format installed.

**Comment by JoergAtGithub** [comment]
This works!

**Comment by Holzhaus** [comment]
> This works!

Great.

**Comment by Holzhaus** [comment]
I rebased on latest 2.3. @JoergAtGithub Can you confirm that with this PR merged, all hooks work as expected on WIndows?

**Comment by Holzhaus** [comment]
Btw, this solution is nice anyway, because we can pin a specific clang-format version and don't require contributors to install clang-format themselves anymore.

**Comment by JoergAtGithub** [comment]
I merged this PR into Main and codespell and clang-format work both.

**Comment by ywwg** [comment]
do we need this in 2.3 or is it sufficient to have it in main?

**Comment by Holzhaus** [comment]
Depends on whether we want windows developers being able to contribute to the `2.3` branch or just `main`.

File: .clang-format [hunk 1]
```diff
@@ -5,9 +5,10 @@ IndentWidth: 4
 TabWidth: 8
 UseTab: Never
 # A ColumnLimit > 0 causes clang-format to unbreaks all short lines,
-# which is undesired here.
-# If the line length exceeds 100, "ColumnLimit:  80" is used in scripts/line-length.py
-ColumnLimit: 0
+# but unfortunately there is no config option to stop it from doing that, and
+# it's a small price to pay for avoiding brittle and complex custom formatting
```
Comment (daschuer) [inline .clang-format:9]: This was IMHO a very good compromise to get a real good formatting. I really don't want to loose that feature. 
What are the alternatives?

**Review by daschuer — CHANGES_REQUESTED** [review]

**Review by daschuer — CHANGES_REQUESTED** [review]
Here are alternatives discussed but without a solution 
https://stackoverflow.com/questions/33656800/clang-format-line-breaks

**Comment by Holzhaus** [comment]
> This was IMHO a very good compromise to get a real good formatting. I really don't want to loose that feature.
> What are the alternatives?

I don't see any. And I'm not sure that the old solution with the custom wrapper script was that good, because not only does it not work on Windows, it also has a bunch of other issues:

It quite complex and requires hundreds of line of custom python code and executing clang-format multiple times on the same source file.

A pretty big culprit is that people that use IDEs with built-in support for clang-format are confused, because a `.clang-format` config exists, gets picked up by the IDE but does not actually describe how the code should look. It conflicts with the formatting we actually want to achieve. This has happened multiple times already. 

I agree that clang-format's unwrapping of short lines to match the column limit is a nuisance (albeit a small one), but there are many projects that use clang-format and if all of those can live with it, so can we.

**Comment by uklotzde** [comment]
If the readability of your code depends on formatting and line length then the code should better be rewritten instead of tuning the formatter.

**Comment by daschuer** [comment]
What is the reason that the scripts do not run under windows?

**Comment by JoergAtGithub** [comment]
This script is not a good citizen - it doesn't respect several environment settings:

1. If modern IDEs like VisualStudio detect the the `.clang-forma`t and formats the code while typing, but the script is not triggered - the script must be triggered by clang-format, and not the other way around
2. The script doesn't use the Python interpreter path, provided by pre-commit. Instead it tries to detect the Python interpreter itself - which doesn't work on Windows and result in a random Python installation on other systems
3. The script doesn't use the clang-format tool provided by pre-commit fails on Windows and use a random version on other systems - which can result in different formatting on CI and local pre-commit runs

**Comment by daschuer** [comment]
OK, I understand. I think we can fix that most of it. I will have a look. 

The IDEs ".clang-format" pickup might be harder to solve. 
I personally like the restrained behavior of the "ColumnLimit: 0" option during editing.
It does nothing wrong, it is only one disabled check.  
This way I can line break whenever I like and clang-format cares for the indentation. 
The pre-commit hook kicks only in on long lines > 100. 

I am curious if to the upstream responds for:  
https://github.com/llvm/llvm-project/issues/58417
With a pinned clang-format version we can use even a pending PR fro this Issue. 
Unfortunately even than we need to tell the IDE where the pinned version is installed, so we can also tell it to use directly pre-commit anyway.

**Comment by daschuer** [comment]
Oh that was easy. I when ahead and created a PR form my tests, sorry for the concurrent PR. 
https://github.com/mixxxdj/mixxx/pull/11004

**Comment by daschuer** [comment]
@Holzhaus can we close this one in favor of https://github.com/mixxxdj/mixxx/pull/11004 this fixes "all" issues with python script on windows. 

Without the script the whole file is formated which is not desired. We want the git-clang-format behaviour. This is where our script originates from.

**Comment by Holzhaus** [comment]
> can we close this one in favor of https://github.com/mixxxdj/mixxx/pull/11004 this fixes "all" issues with python script on windows.

Your PR is still broken in multiple ways that @JoergAtGithub and me pointed out above, e.g. IDE integration is broken because the `.clang-format` config does not match out actual code style, it still requies a complicated wrapper script to archieve non-standard formatting and using `pre-commit run clang-format --all-files` or `pre-commit run --files path/to/file.cpp` is still completely borked and always says everything "passed".

But whatever, I stopped caring.

**Pull request closed without merging**